body.tex

```diff
@@ -1,166 +1,466 @@
 
-\begin{itemize}
-   \item Telescope and site overview and status in 2014:  \cite{2014SPIE.9145E..1AG}
-\end{itemize}
-%------------------------------------------------------------------------------
+\subsection{PostgreSQL performance}
 
-\subsubsection{System Engineering}
+The following described performance issue applies strictly to w\_2019\_38 builds and earlier. The QuantumGraph is created by issuing a very large SQL statement that, effectively, creates a cartesian product between many, of the 15-ish or so, tables in the registry. The results of the large query are then parsed in Python and a slew of many different, small, follow-up queries are issued. This diverse workflow presents a difficult challenge in DB optimization. The Generation 3 Middleware team reports that on average it takes 0.5 to 1.5h to create QuantumGraphs for tract-size DRP workflows with Oracle on NCSA infrastructure. In our tests we were unable to create even the simplest QuantumGraphs, even after doubling of the RDS instance resources, and were forced to abort multiple times after 30+h of execution. The lack of performance did not seem related to hardware limitations, since none of the performance metrics availible showed heavy hardware loads nor didthe problem go away even after significant increase of RDS resources.
 
-\begin{itemize}
-   \item LSST systems engineering: \cite{2014SPIE.9150E..0MC}
-   \item System verification and validation: \cite{2014SPIE.9150E..0NS}
-\end{itemize}
-%
+PostgreSQL, and Oracle, regularly collect statistics on database objects used in queries. These statistics are then used to generate execution plans. These execution plans are then oftentimes cached. This in general results in performance gains. But because Butler generates all SQL dynamically there are no guarantees that a cached, or even pre-prepared execution plan can be executed. In PostgreSQL many of these optimizations then fail.
+
+Specifically what seems to happen with PostgreSQL is that all the views materialize completely on disk. Even when the views are part of larger statements, in which the outter statements have strict constraints on them, the outter constraints do not seem to penetrate to the view statement. It is unclear, from within \poc, whether Oracle DBMS does not suffer from this issue because it has a better SQL statement optimization engine or if this is something more directly tied to how PostgreSQL implements views. 
```

Comment:
Fix "outer", but also leave out the Oracle statement: it doesn't add any information.

Reply:
Fixed "outter". Removed last sentence. Resolved.